common/txmgr/tracker.go

```diff
@@ -99,75 +96,85 @@ func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Start(ctx c
 }
 
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) startInternal(ctx context.Context) (err error) {
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
-
-	tr.ctx, tr.ctxCancel = context.WithCancel(context.Background())
+	tr.initSync.Lock()
+	defer tr.initSync.Unlock()
 
 	if err := tr.setEnabledAddresses(ctx); err != nil {
 		return fmt.Errorf("failed to set enabled addresses: %w", err)
 	}
-	tr.lggr.Info("Enabled addresses set")
-
-	if err := tr.trackAbandonedTxes(ctx); err != nil {
-		return fmt.Errorf("failed to track abandoned txes: %w", err)
-	}
-
-	tr.isStarted = true
-	if len(tr.txCache) == 0 {
-		tr.lggr.Info("no abandoned txes found, skipping runLoop")
-		return nil
-	}
+	tr.lggr.Infof("enabled addresses set for chainID %v", tr.chainID)
 
-	tr.lggr.Infof("%d abandoned txes found, starting runLoop", len(tr.txCache))
+	tr.chStop = make(chan struct{})
 	tr.wg.Add(1)
-	go tr.runLoop()
+	go tr.runLoop(tr.chStop.NewCtx())
+	tr.isStarted = true
 	return nil
 }
 
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Close() error {
-	tr.lock.Lock()
-	defer tr.lock.Unlock()
 	return tr.StopOnce("Tracker", func() error {
 		return tr.closeInternal()
 	})
 }
 
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) closeInternal() error {
+	tr.initSync.Lock()
+	defer tr.initSync.Unlock()
+
 	tr.lggr.Info("stopping tracker")
 	if !tr.isStarted {
-		return fmt.Errorf("tracker not started")
+		return fmt.Errorf("tracker is not started: %w", services.ErrAlreadyStopped)
 	}
-	tr.ctxCancel()
+
+	close(tr.chStop)
 	tr.wg.Wait()
 	tr.isStarted = false
 	return nil
 }
 
-func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) runLoop() {
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) runLoop(ctx context.Context, cancel context.CancelFunc) {
 	defer tr.wg.Done()
+	defer cancel()
+
+	if err := tr.trackAbandonedTxes(ctx); err != nil {
+		tr.lggr.Errorf("failed to track abandoned txes: %v", err)
+		return
+	}
+	if err := tr.handleTxesByState(ctx, 0); err != nil {
+		tr.lggr.Errorf("failed to handle txes by state: %v", err)
+		return
+	}
+	if tr.AbandonedTxCount() == 0 {
```

Comment:
Don't we want the tracker to run in the background in case new addresses get abandoned? Is it only during startup?

Reply:
We only know if an address is abandoned after node startup, and in the case that keys are changed while the node is running a restart will be triggered and the tracker will then pick up the new abandoned txes.